Added fix to Camera.cpp not correctly calculating the local norm

victoronline opened the pull request:
Added fix to half pixel issue in Camera.cpp - ISIS Issue [#4018](https://github.com/USGS-Astrogeology/ISIS3/issues/4018), whereas DBL_MIN is not operating as expected, causing an error later in the code, when integer rounding occurs. This erroneously returns the next/previous framelet. 

## Description
Currently the DBL_MIN is not operating as expected. The functionality that is intended is:
Calculate normal by grabbing elevation values at the edges of the pixel in question.

What is actually happening:
The values gathered are still not from the same pixel. It appears as though the DBL_MIN is not subtracting more than the rounding error, and this is essentially returning the same value as if the code was:

      QList< QPair< double, double > > surroundingPoints;
      surroundingPoints.append(qMakePair(samp, line - 0.5));
      surroundingPoints.append(qMakePair(samp, line + 0.5));
      surroundingPoints.append(qMakePair(samp - 0.5, line));
      surroundingPoints.append(qMakePair(samp + 0.5, line));

This change was made to version 3.10.2. 

## Motivation and Context
This is detrimental when calculating photometric angles for push-broom and push-frame cameras, because 1.0 could be the pixel center, 0.5 is the edge of pixel 1, but 1.5 is the edge of pixel 2. This does not work well when a push-broom imager is flying in a reverse direction, or for frame boundaries of a push-frame camera. The calculation is using a completely wrong location on the body.


## How Has This Been Tested?
The LROC team has tested the fix. Will provide details in comments below but a test (GTest) has not been created for this yet.

## Screenshots (if appropriate):
Please see issue. Several screenshots and examples have been included.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ ] My code follows the code style of this project. -->
- [X] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [X] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [X] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

victoronline committed:
Added fix to half pixel issue in Camera.cpp where pexels on edge of framelet are rounding incorrectly due to the integer cast in the detector map - ISIS [issue 4018]

victoronline committed:
removed commented code and updated header file with change history

victoronline committed:
fixed ypo

victoronline committed:
fixed text formatting in Camera.h history

jessemapel commented on isis/src/base/objs/Camera/Camera.cpp:
```diff
@@ -1540,9 +1540,9 @@ namespace Isis {
       QList< QPair< double, double > > surroundingPoints;
       surroundingPoints.append(qMakePair(samp, line - 0.5));
-      surroundingPoints.append(qMakePair(samp, line + 0.5 - DBL_MIN));
+			surroundingPoints.append(qMakePair(samp, std::nexttoward(line + 0.5, line)));
```
Looks like a weird spacing issue is happening here. Check your tabs vs spaces in your editor.

acpaquette commented:
@victoronline This should be picked up by a unit test so I don't think a gtest will be necessary. I'm concerned that this will have a pretty big impact since this is a change to the base camera

jessemapel reviewed (commented):
Change looks good. I'm trying to think of the best way to test this. We could make a camera model that just records its framelet and test this function on the edge of a framelet and then check which framelet it sets to.

jessemapel commented:
Fixes #4018

jlaura commented:
@victoronline Awesome to see the PR come in! 👍 🎉 

@Kelvinrr @AustinSanders @acpaquette @kaitlyndlee Can we get this slotted into the support sprint this week or next to assist getting a GTest written?

What is everyone thinking for getting this released post merge? We should be able to cherry pick this into dev after a merge into 3.10 and it will then be released during the normal release process next month (4.3.0_RC and then 4.3.0). Victor, are you all running this change on your own 3.10 custom build?

scsides commented on isis/src/base/objs/Camera/Camera.cpp:
```diff
@@ -1540,9 +1540,9 @@ namespace Isis {
+			surroundingPoints.append(qMakePair(samp, std::nexttoward(line + 0.5, line)));
       surroundingPoints.append(qMakePair(samp - 0.5, line));
-      surroundingPoints.append(qMakePair(samp + 0.5 - DBL_MIN, line));
+      surroundingPoints.append(qMakePair(std::nexttoward(samp + 0.5, samp), line));
```
These two changes make sense, and will fix the problem for phocube. If you want a slightly more accurate answer, you should adjust line 1542 and 1544 in a similar way as you did for 1543 and 1545. This will keep the requested pixel (line/sample) centered between the four surrounding pixels.

Although this does fix this function when phocube uses it with integer sample/line values, it does not fix it when other applications use non-integer sample/line values. We have created a new issue #4256 to address what we believe to be a fix for the other cases.

As far as testing this change goes: We do not believe it is reasonable to test this change under the current unittest environment. This would be much more reasonable once the camera test has been converted to gtest. Given that, this PR should be mergeable after addressing the current comments.

victoronline replied:
Working to implement changes. Thanks.

jessemapel commented:
@victoronline Any update on this? I'll echo Stuart and say that this is mergable into dev once his comments have been addressed.

#4256 can be addressed in a separate PR to fix this for non-integer pixel values.

victoronline committed:
Adjusted lines 1542, 1544 to keep pixel(line/sample) centered.

Adjusted line 1542 and 1544 in a similar way as was done for lines 1543 and 1545 in order to keep the requested pixel (line/sample) centered between the four surrounding pixels.

victoronline reviewed (commented):
Changes made.

jessemapel commented:
@victoronline 3.10 is not an active branch any more, can you change this PR to go into dev?

scsides reviewed (changes requested):
All the code changes look good.

The gtest submodule changes (266 files) needs to be fixed. We don't want to include them in this PR. Can you try to re-checkout gtest to revert these changes?

victoronline commented:
Since 3.10 is no longer an active branch, but the version we need changes made to, what does that mean if putting into dev? Will this change only be reflected in the latest ISIS version?

jessemapel commented:
Yes, this change will only be available in the newest release. You can also cherry-pick the PR from this commit back onto the 4.4 branch and then a 4.4.1 release can be pulled without some of the other new functionality, but we only go back 1 feature release right now.

victoronline commented:
Reopening due to removing local branch in error.

victoronline commented:
Submitted change to dev branch via PR [#4340](https://github.com/USGS-Astrogeology/ISIS3/pull/4340) since branch for this PR is no longer active (3.10.2). Closing this PR.

scsides reviewed (commented):
Woops, missed the camera.h history comment.
